homeassistant/components/thethingsnetwork/config_flow.py

```diff
@@ -0,0 +1,116 @@
+    TextSelectorConfig,
+    TextSelectorType,
+)
+
+from .const import CONF_API_KEY, CONF_APP_ID, CONF_HOSTNAME, DOMAIN, TTN_API_HOSTNAME
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TTNFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize flow."""
+        self._reauth_entry: ConfigEntry | None = None
+
+    async def async_step_user(
+        self, user_input: Mapping[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """User initiated config flow."""
+
+        errors = {}
+        if user_input is not None:
+            connection_error = None
+            try:
+                client = TTNClient(
+                    user_input[CONF_HOSTNAME],
+                    user_input[CONF_APP_ID],
+                    user_input[CONF_API_KEY],
+                    0,
+                )
+                await client.fetch_data()
+            except TTNAuthError:
+                # Raising ConfigEntryAuthFailed will cancel future updates
+                # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+                _LOGGER.error("TTNAuthError")
+                connection_error = "invalid_auth"
+            except Exception as err:  # pylint: disable=broad-exception-caught
+                _LOGGER.error(err)
+                _LOGGER.error(traceback.format_exc())
+                connection_error = "unknown"
+
+            if connection_error:
+                errors["base"] = connection_error
+            else:
+                # Create entry
+                if self._reauth_entry:
+                    return self.async_update_reload_and_abort(
+                        self._reauth_entry,
+                        data=user_input,
+                        reason="reauth_successful",
+                    )
+                if not self.unique_id:
+                    await self.async_set_unique_id(user_input[CONF_APP_ID])
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(
+                    title=str(user_input[CONF_APP_ID]),
+                    data=user_input,
+                )
+
+        # Show form for user to provide settings
+        if not user_input:
+            if self._reauth_entry:
+                user_input = self._reauth_entry.data
+            else:
+                user_input = {CONF_HOSTNAME: TTN_API_HOSTNAME}
+
+        schema = vol.Schema(
+            {
+                vol.Required(CONF_HOSTNAME, default=user_input[CONF_HOSTNAME]): str,
+                vol.Required(CONF_APP_ID, default=user_input.get(CONF_APP_ID)): str,
+                vol.Required(
+                    CONF_API_KEY, default=user_input.get(CONF_API_KEY)
+                ): TextSelector(
+                    TextSelectorConfig(
+                        type=TextSelectorType.PASSWORD, autocomplete="api_key"
+                    )
+                ),
+            }
+        )
```

Comment:
you can use `self.add_suggested_values_to_schema` for this

Reply:
`add_suggested_values_to_schema` seems to add suggest values and not defaults - so I get errors because the required values are not set.

Basically the patter I am trying to do is:
1. the user enters is input in the form or this is a reauth flow
2. the connection fails
3. the user is presented the form again with the values entered in 1 so they can be checked and corrected.
    - this is where the default values coming from the old user_input are needed

Is there an `add_default_values_to_schema` ?

Reply:
Now I think of it, whenever someone reauths, what is invalid? Probably the API key, so we'd rather only show the form with the API key. That does not have to be prefilled as that is invalid.

How does the frontend show the suggested values, I always understood that it does show it like the default ones

Reply:
in the case of ttn is possible to rename the app_id. Even the host might change (not sure about this last one). The API error does not tell more than the API does not work. This is why I thought about using the same schema as initially.

According to the [documentation](https://developers.home-assistant.io/docs/data_entry_flow_index#show-form) with suggestion the user could let the field empty which is not intended as the 3 fields are required.

Also, to give it a try - I changed the Required values to optional - this allowed the schema validation to work but the suggestions were not added in the testcases - this might only work with the UI?

Should I (in another PR) contribute a `add_default_values_to_schema` ? It would be the same logic, just adding defaults instead of optional suggestions?

Reply:
As reference, this is what I tested:
![image](https://github.com/home-assistant/core/assets/4406403/719e5505-25d9-4718-b36a-e91ecbb38e1c)

Reply:
If the app id is changeable, the app id is not a good unique id to use. When can this change?

You can try submitting that, no clue if it will be accepted, I think we usually use suggested values in these cases (in new cases at least)

Reply:
investigating further - `add_suggested_values_to_schema` works with required when using the UI. So it is only the testcases that need to be adjusted to receive suggestions instead of defaults.

Reply:
added a helper function in the testcase at https://github.com/angelnu/core/blob/thethingsnetwork_v3/tests/components/thethingsnetwork/__init__.py#L4-L20 so that the suggested values can be tested.

Inpired in the testcases of the `obihai` integration.